### ISSUE TITLE
Support immutable types?

### ISSUE THREAD

**Issue by ebalaban** [sisl/POMDPs.jl#53]
I think that using immutable types for states, actions, and observations could result in noticeable performance improvements for some (perhaps even all) problem/solver combinations. We would have to change at least a few things in the API if we want to support such types, however. Here are some of these things:

- create_state(), create_observation(), and create_action() functions would become somewhat meaningless and all the places in one's code that currently rely on being able to modify state, action, and observation objects would need to be redone.

- rand!(rng::AbstractRNG, sample::Any, d::AbstractDistribution) and rand!(rng::AbstractRNG, sample::Any, space::AbstractSpace) would need to be modified to allocate and return immutable objects. Doing something like rand!(rng::AbstractRNG, state = create_state(pomdp), d::TransitionDistribution) wont work, unfortunately (see above).

Since introducing support for immutable types may be rather disruptive, should we do it? If so, should we limit it to only certain types (e.g. states, actions, and observations) or not? For instance, it might be possible to make distribution types immutable, but that will likely not result in as much of a performance payoff (compared to states, actions, and observations).

**Comment by mykelk** [comment]
The create functions can just return some default immutable. The only real change is rand, and that change would really just make it consistent with the other functions, right? I don't think this is a major change other than to modify the instances of rand in our code... Unless I'm missing something.

**Comment by ebalaban** [comment]
Well, that default immutable won't be modifiable (and that was the main reason for having create_* functions, aside of just being able to get the type of the returned object). Am I missing something here?

**Comment by mykelk** [comment]
Right, they won't be modifiable, but that's okay. The problem implementer won't try to modify it when it is passed into rand, etc.

**Comment by ebalaban** [comment]
Ok, sure, I can see how the value assigned to the input argument of rand can just be ignored if the problem writer expects an immutable,  but what about all the places in the solvers where create_* functions are used for preallocation? And all the places where the solver writer passes a variable into rand expecting it to be modified, rather than relying on doing x = rand()? Will we need to start distinguishing between solvers that support immutable types and those that don't?

**Comment by etotheipluspi** [comment]
@ebalaban in your performance example you are comparing creating new immutable types to new mutable types at every iteration. Something that more closely resembles our solvers would just modify the value of the mutable type and use it in a computation. I originally thought that this would lead to nearly the same performance, but I ran a few test and was surprised to find a huge difference in compute time still. 

Here is the code: 

```julia
abstract AbstractState

type StateT <: AbstractState
    x::Int64
    y::Int64
end

immutable StateI <: AbstractState
    x::Int64
    y::Int64
end

function test(n::Int64)
    println("Mutable:"); t1 = @time test1(n)
    println("Immutable:"); t2 = @time test2(n)
    println("Baseline:"); t3 = @time test3(n)
    @assert t1 == t2 == t3
    return nothing
end


function test1(n::Int64)
    tot = 0
    s = StateT(0,0)
    for i=1:n,j=1:n
        s.x = i
        s.y = j
        tot += sum_state(s)
    end
    return tot
end

function test2(n::Int64)
    tot = 0
    for i=1:n,j=1:n
        tot += sum_state(StateI(i,j))
    end
    return tot
end

function test3(n::Int64)
    tot = 0
    for i=1:n,j=1:n
        tot += (i+j)
    end
    return tot
end

function sum_state(s::AbstractState)
    return s.x+s.y
end                                                                                                                  
```

These are the results:
```
julia> test(50_000)
Mutable:
  2.574183 seconds (1 allocation: 32 bytes)
Immutable:
  0.000026 seconds
Baseline:
  0.000027 seconds
```

You can look at what Julia compiles this down to by running
```@code_llvm test1(10_000)``` and ```@code_llvm test2(10_000)```. My guess is Julia compiles away a lot of the overhead when working with immutable types like this (accessing fields for example). There could also be some fundamental issue I'm missing here. 

I have two things:
Does anyone have any idea why the performance gap between mutable and immutable types is so large here even though we are allocating almost no memory?

If this is truly the difference in performance we should support immutable types.

**Comment by ebalaban** [comment]
@etotheipluspi, yes, that's a good example. I was just trying something quick and dirty yesterday (although it did also show how Julia pads allocated memory for mutable types to give them "room to grow", I am guessing). I ran your code on my end and got similar results.

**Comment by mykelk** [comment]
Yes, this would require changing existing solvers to support the new rand. We should see what the others think. I don't think the calls to create have to change.

**Comment by etotheipluspi** [comment]
@mykelk do you have a sense of why the above example has such a huge performance difference between mutable and immutable types?

**Comment by etotheipluspi** [comment]
We should also consider how this will affect states/actions/observations that are represented as arrays. For example:

```julia
immutable State
    vel_vector::Array{Float64}
    pos_vector::Array{Float64} 
end
```

While the above can be modified to

```julia
immutable
    x_vel::Float64
    y_vel::Float64
    x_pos::Float64
    y_pos::Float64
end
```
It will make defining states with a large number of variables more cumbersome.

**Comment by ebalaban** [comment]
@mykelk, yes, I suppose the calls to create() functions won't have to change, they just won't be terribly useful in immutable cases. As for rand(), the solvers will all need to use x = rand() to support both mutable and immutable types, but I guess there is no particular harm in leaving the optional "sample" argument.

**Comment by mykelk** [comment]
Maybe [this](http://julialang.org/blog/2013/03/efficient-aggregates/) helps with the understanding?

**Comment by ebalaban** [comment]
That's a good post. The difference in performance is pretty impressive.

**Comment by ebalaban** [comment]
@etotheipluspi, maybe I misunderstood your comment above, but you can still have array types as fields in an immutable type if you want, they just won't be very efficient.

Out of curiosity, I tried the following (with Int64 instead of Float64):
```julia
abstract ArrayState
abstract FlatState

type MutableArrayState <: ArrayState
    vel_vector::Array{Int64}
    pos_vector::Array{Int64} 
end

immutable ImmutableArrayState <: ArrayState
    vel_vector::Array{Int64}
    pos_vector::Array{Int64} 
end

type MutableFlatState <: FlatState
    x_vel::Int64
    y_vel::Int64
    x_pos::Int64
    y_pos::Int64
end

immutable ImmutableFlatState <: FlatState
    x_vel::Int64
    y_vel::Int64
    x_pos::Int64
    y_pos::Int64
end

function test(n::Int64)
    println("Mutable type with arrays:"); t1 = @time test_mutable_array(n)
    println("Immutable type with arrays:"); t2 = @time test_immutable_array(n)
    println("Mutable flat type:"); t3 = @time test_mutable_flat(n)    
    println("Immutable flat type:"); t4 = @time test_immutable_flat(n)
    println("Baseline:"); t5 = @time test_baseline(n)
    @assert t1 == t2 == t3 == t4 == t5
    return nothing
end

function test_mutable_array(n::Int64)
    tot = 0
    s = MutableArrayState([0,0],[0,0])
    for i=1:n,j=1:n
        s.vel_vector[1]=i
        s.vel_vector[2]=j
        s.pos_vector[1]=j
        s.pos_vector[2]=i        
        tot += sum_state(s)
    end
    return tot
end

function test_immutable_array(n::Int64)
    tot = 0
    for i=1:n,j=1:n
        s = ImmutableArrayState([i,j],[j,i])
        tot += sum_state(s)
    end
    return tot
end

function test_mutable_flat(n::Int64)
    tot = 0
    s = MutableFlatState(0,0,0,0)
    for i=1:n,j=1:n
        s.x_vel = i
        s.y_vel = j
        s.x_pos = j
        s.y_pos = i
        tot += sum_state(s)
    end
    return tot
end

function test_immutable_flat(n::Int64)
    tot = 0
    for i=1:n,j=1:n
        tot += sum_state(ImmutableFlatState(i,j,j,i))
    end
    return tot
end

function test_baseline(n::Int64)
    tot = 0
    for i=1:n,j=1:n
        tot += (i+j+j+i)
    end
    return tot
end

function sum_state(s::ArrayState)
    return sum(s.vel_vector)+sum(s.pos_vector)
end

function sum_state(s::FlatState)
    return s.x_vel+s.y_vel+s.x_pos+s.y_pos
end
```

Here are the results:
Mutable type with arrays:
  4.714813 seconds (49.74 M allocations: 758.965 MB, 1.24% gc time)
Immutable type with arrays:
  7.575971 seconds (214.52 M allocations: 6.549 GB, 6.66% gc time)
Mutable flat type:
  0.037028 seconds (1 allocation: 48 bytes)
Immutable flat type:
  0.000002 seconds
Baseline:
  0.000002 seconds

So, having an immutable type with just arrays does not buy you anything in terms of performance (as one would expect...), but it works just fine.

**Comment by etotheipluspi** [comment]
@ebalaban right, because when you have an array in an immutable type, you have to allocate memory every time you create your type. With mutable types, you can pre-allocate which is more efficient. Sorry I wasn't more clear, that's what I meant. 

Also, I'm not sure that post by Jeff explain the discrepancy in performance. I understand why immutable are fast (in fact as fast as doing the addition operation with the baseline). What I don't understand is why mutable types are so slow when doing simple operations.

**Comment by ebalaban** [comment]
@mykelk, is this roughly how you envisioned the new rand!() working? I don't think we should make `sample` optional, as a POMDP object would be required to create the default value. I think that `s = rand!(rng, s, d)` syntax is not so bad, but if you had something more elegant in mind, please let me know.

```julia
abstract AbstractState
abstract AbstractDistribution
abstract POMDP

type MutableState <: AbstractState
    x::Int64
end

immutable ImmutableState <: AbstractState
    x::Int64
end

type MutablePOMDP <: POMDP
    rng::AbstractRNG 
end

type ImmutablePOMDP <: POMDP
    rng::AbstractRNG
end

type SimpleRNG <: AbstractRNG
    seed::Int64
end

type SimpleDist <: AbstractDistribution
    mean::Int64
end

create_state(::MutablePOMDP) = MutableState(-1)
create_state(::ImmutablePOMDP) = ImmutableState(-1)
create_dist(::POMDP) = SimpleDist(20)

function test()
    rng = SimpleRNG(5)
    m_pomdp = MutablePOMDP(rng)
    i_pomdp = ImmutablePOMDP(rng)    
    
    simulate(m_pomdp)
    simulate(i_pomdp)
    return nothing
end
    
function rand!(rng::AbstractRNG, sample::MutableState, dist::AbstractDistribution)
    sample.x = dist.mean*rng.seed    
    return sample
end

function rand!(rng::AbstractRNG, ::ImmutableState, dist::AbstractDistribution)
    return ImmutableState(dist.mean*rng.seed)
end

function simulate(pomdp::POMDP)
    d = create_dist(pomdp)
    s = create_state(pomdp)
    
    s = rand!(pomdp.rng, s, d)
    println("$(typeof(pomdp)): new s = $s")
end
```

**Comment by mykelk** [comment]
Yes, that's exactly what I had in mind.

**Comment by ebalaban** [comment]
Ok, good, we are on the same page then. If we do decide to go this way, we may want to discuss whether `rand()` or `rand!()` is less confusing, but that's a minor issue.

**Comment by mykelk** [comment]
I think we want it with ! because we mutate the rng, right? Or is rand a special case?

**Comment by etotheipluspi** [comment]
Only the state gets mutated in ```rand!```. At least that's how I've been doing it.

**Comment by mykelk** [comment]
The random number generator state gets changed, right?

**Comment by ebalaban** [comment]
In Base both rand and rand! can take (and mutate) rng as an optional argument. The difference between the two is that rand! mutates a sample variable/array in place, whereas rand allocates and returns a new variable/array. Ours would do one or the other depending on the sample type, so we could go either way. I think rand() would be a little less confusing, given that we'd almost always write it as s=rand(rng,s,d), but I am fine with leaving it as rand!() as well.

**Comment by zsunberg** [comment]
Yeah, I think that to support immutables, we would have to diverge from
Base a little bit. `rand(rng, distribution, state)` where the state
argument is optional seems consistent with the rest of the interface

**Comment by zsunberg** [comment]
by "the interface" I mean our POMDPs.jl interface

**Comment by mykelk** [comment]
Okay, I'm fine with or without the !.

**Comment by ebalaban** [comment]
`rand()` naming issue aside, @etotheipluspi and @zsunberg, are you guys ok with introducing support for immutables? I don't think it's going to be as disruptive as I imagined at first. If so, do we limit this support to states/actions/observations or do we allow distributions be immutable also, for example? `transition` and `observation` would then need to change similarly to `rand()`. Where do we draw the line?

**Comment by etotheipluspi** [comment]
I am ok with this as long as the only change is going from ```rand!``` to ```rand```. 

If we were to support immutable distributions, would ```transition``` and ```observation``` need to change?

**Comment by mykelk** [comment]
The `transition` and `observation` functions already return the things they (potentially) modify. There is nothing to change there.

**Comment by ebalaban** [comment]
You are right, `transition` and `observation` are actually fine in their current form. So I guess we don't have to modify or stipulate anything then. Even if someone wants to make spaces immutable, they should be able to do so as well. So, I guess we need to decide whether to rename `rand!()` or leave it alone. My slight preference is for `rand()`, to bring it in line with the rest of the API, but I'm also fine with leaving it as is.

**Comment by mykelk** [comment]
I also have a slight preference for `rand()`.

**Comment by etotheipluspi** [comment]
I'm ok with ```rand()``` as well.

**Comment by ebalaban** [comment]
@zsunberg, from your comment above I am assuming that you are ok, in principle, with supporting immutables and renaming from `rand!()` to `rand()`. Making `state` optional is a bit problematic because we'd need to supply a POMDP object to `create_state()`. If you agree, I'll go ahead and make the change.

**Comment by zsunberg** [comment]
Yeah, I am ok with it

**Comment by ebalaban** [comment]
Great, thanks. Implemented with 07ab7dd.

**Comment by zsunberg** [comment]
Sorry, I was on a different page from y'all. I thought we would be using `rand(rng, distribution, state)` to match the rest of the interface, but it is in the package as `rand(rng, state, distribution)`. I feel pretty strongly that it should be the other way for consistency's sake (the only reason I can imagine to keep (rng, state, sistribution) would be for consistency with the former `rand!`, but this function doesn't exist anymore, so that doesn't seem like a good reason)

**Comment by ebalaban** [comment]
I just did a minimally disruptive change from 'rand!()` (which, I am guessing, followed the format of `Base.rand()`). I am fine with going to `rand(rng, distribution, state)` if the others agree.

**Comment by mykelk** [comment]
I'm okay with that.

**Comment by ebalaban** [comment]
@etotheipluspi, are you ok with the change @zsunberg proposed? If you are, I'll go ahead and change the arg order.

**Comment by etotheipluspi** [comment]
yep go ahead!

**Comment by ebalaban** [comment]
Great, thank you sir. Implemented with ae832b1 and 79b8869. The (now) last argument was called 'state' in some places and 'sample' in others. I standardized to 'sample', as it is more general.
